asllib/doc/ASLSemanticsReference.tex

```diff
@@ -19,160 +283,1634 @@
+%
+We note that our semantics always maintains the same static environment.
+
+\subsection{Execution Graphs}
+
+The concurrent semantics of an ASL specification utilizes \herd\ \emph{execution graphs}~\cite{AlglaveDGHM21},
+which track the read/write effects over variables, yielded by the sequential semantics, and the \emph{ordering constraints}
+between those effects.
+
+Formally, an execution graph $\xgraph = (N^\xgraph, E^\xgraph, O^\xgraph) \in \XGraphs$
+is defined via a set of \emph{nodes} ($N^\xgraph$), \emph{edges} ($E^\xgraph$), and \emph{output nodes} ($O^\xgraph$):
+\[
+\begin{array}{rcl}
+  \XGraphs   	    &\triangleq& \pow{\Nodes} \times \pow{\Nodes \times \Nodes \times \Labels} \times \Nodes  \\
+  \Nodes            &\triangleq& \N \times \{ \Read, \Write \} \times \identifier \\
+  \Labels           &\triangleq& \{ \asldata, \aslctrl, \aslpo \}
+\end{array}
+\]
+
+Nodes represent unique read/write effects. Formally, a node $(u,l,\id)$ associates a unique instance counter $u$
+to an \emph{ordering label} $l$, which specifies whether it represents a read or write effect to a variable named $\id$.
+
+An edges represents an ordering constraint between two effects, which can be one of the following:
+\begin{description}
+\item[$\asldata$] Represents a \emph{data dependency}.
+That is, when a write effect reads/write over the data written by another write effect.
+\item[$\aslctrl$] Represents a \emph{control dependency}.
+That is, when a read effect to a variable determines the flow of control (e.g., which condition of a branch is taken),
+which then leads to another read/write effect.
+\item[$\aslpo$] Represents a \emph{program order}.
+That is, when two effects are defined to be ordered according to the sequential semantics of the program.
+\end{description}
+
+An execution graph is \emph{well-formed} if all nodes have unique instance counters, edges connect graph nodes,
+and the output nodes are contained in the set of nodes:
+\[
+  \begin{array}{rcl}
+  \forall n, n' \in N^\xgraph &.& n=(u,l,\id) \wedge n=(u',l',\id') \;\Rightarrow\; u \neq u' \\
+  \forall e \in E^\xgraph &.& e=(n, n', l) \;\Rightarrow\; n, n' \in N^\xgraph \\
+  & & O^\xgraph \subseteq N^\xgraph \enspace.
+  \end{array}
+\]
+
+We denote the empty execution graph $\emptygraph \triangleq (\emptyset, \emptyset, \emptyset)$.
+%
+We define the following functions, which return an execution graph that represents a single read/write effect to a variable \texttt{x}:
+\[
+  \begin{array}{rclll}
+    \WriteEffect(\texttt{x}) &\triangleq& (\{n\}, \emptyset, \{n\}) & \text{where } n=(u, \Write, \texttt{x}), & u\in\N \text{ is fresh}\\
+    \ReadEffect(\texttt{x})  &\triangleq& (\{n\}, \emptyset, \{n\}) & \text{where } n=(u, \Read, \texttt{x}),   & u\in\N \text{ is fresh}\\
+  \end{array}
+\]
+
+We also define two ways to combine execution graphs --- \emph{unordered combination} and \emph{ordered combination}.
```

Comment:
combine -> compose
unordered combination -> parallel composition
ordered combination -> are there not several ways to order the graphs? namely, at least data and ctrl? why do you only have one?

Reply:
I'll clarify that the ordered combination is parameterized by a label:
> ... and \emph{ordered combination} with a given label.